### PR TITLE
8240908: RetransformClass does not know about MethodParameters attribute

### PR DESCRIPTION
**Pull request by alexmenkov**
Changes:
- ClassFileReconstituter is updated to restore "MethodParameters" attribute;
- handling of the attribute in VM_RedefineClasses is moved to be consistent with other code (like local variable table);
- copied ClassTransformer class (from test/jdk/com/sun/jdi/lib/jdb) to /test/lib as it's used by tests from hotspot and jdk (and also by test from Valhalla repo); Will file a follow up issues to updates tests and remove the class from test/jdk/com/sun/jdi/lib/jdb

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8240908](https://bugs.openjdk.java.net/browse/JDK-8240908): RetransformClass does not know about MethodParameters attribute

### Reviewers
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to 0f80f7657b93aed817c3e74e5e2a0ca5e7bb85dc
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to 0f80f7657b93aed817c3e74e5e2a0ca5e7bb85dc

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7180/head:pull/7180` \
`$ git checkout pull/7180`

Update a local copy of the PR: \
`$ git checkout pull/7180` \
`$ git pull https://git.openjdk.java.net/jdk pull/7180/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7180`

View PR using the GUI difftool: \
`$ git pr show -t 7180`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7180.diff">https://git.openjdk.java.net/jdk/pull/7180.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/prims/jvmtiRedefineClasses.cpp [hunk 1]
```diff
@@ -3694,6 +3679,19 @@ void VM_RedefineClasses::set_new_constant_pool(
       } // end for each local variable table entry
     } // end if there are local variable table entries
 
+    // Update constant pool indices in the method's method_parameters.
+    int mp_length = method->method_parameters_length();
+    if (mp_length >= 0) {
```
Comment (sspitsyn) [inline src/hotspot/share/prims/jvmtiRedefineClasses.cpp:3684]: This line is better to be:
3684       if (mp_length > 0) {

Comment (alexmenkov) [reply src/hotspot/share/prims/jvmtiRedefineClasses.cpp]: Fixed.

**Comment by plummercj** [comment]
The changes look good except I haven't look at the new test that closely. It really could use a high level explanation of how the test works, plus some embedded comments. Too much to figure out by just looking at the code. Also, is there an existing test it was based on? Seems like a lot of code just for testing of a single attribute.

File: test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import jdk.test.lib.JDKToolLauncher;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.util.ClassTransformer;
+
+class MethodParametersTarget {
```
Comment (sspitsyn) [inline test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java:54]: The test has a lack of comments, so it is not easy to read and understand it.
Could you look at it and consider adding some comments where it makes sense to add?
For instance, some explanation is needed on how `@1` with tags `commentout` and `uncomment` are used.
It is not convenient to go deep into the `jdk.test.lib.util.ClassTransformer` to get such answers.

**Comment by alexmenkov** [comment]
> The changes look good except I haven't look at the new test that closely. It really could use a high level explanation of how the test works, plus some embedded comments. Too much to figure out by just looking at the code. Also, is there an existing test it was based on? Seems like a lot of code just for testing of a single attribute.

Many tests in test/jdk/java/lang/instrument are similar - they extends ATransformerManagementTestCase, but this are regression tests for specific bugs. I think it's better to have simple testcase for this issue too.

**Comment by alexmenkov** [comment]
Simplified the test (a lot of code was useful during development/debugging, but now don't have much value).

File: test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java (hunk 2 above)
Comment (alexmenkov) [reply test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java:58]: Done.

File: test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+import jdk.test.lib.JDKToolLauncher;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.util.ClassTransformer;
+
+/*
+ * The test verifies Instrumentation.retransformClasses() (and JVMTI function RetransformClasses)
+ * correctly handles MethodParameter attribute:
+ * - classfile bytes passed to transformers (and JVMTI ClassFileLoadHook event callback) contain the attribute;
+ * - the attribute is updated.
```
Comment (plummercj) [inline test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java:54]: I think you could use a bit more than this. Basically summarize the 3 test cases in doRunTests().

File: src/hotspot/share/prims/jvmtiRedefineClasses.cpp [hunk 4]
```diff
@@ -3694,6 +3679,19 @@ void VM_RedefineClasses::set_new_constant_pool(
...
+    int mp_length = method->method_parameters_length();
+    if (mp_length > 0) {
+        MethodParametersElement* elem = method->method_parameters_start();
+        for (int j = 0; j < mp_length; j++) {
+            const int cp_index = elem[j].name_cp_index;
+            const int new_cp_index = find_new_index(cp_index);
+            if (new_cp_index != 0) {
+                elem[j].name_cp_index = new_cp_index;
+            }
+        }
```
Comment (plummercj) [inline src/hotspot/share/prims/jvmtiRedefineClasses.cpp:3692]: Indentation should be 2, not 4.

**Review by plummercj — CHANGES_REQUESTED** [review]
Thanks for adding comments and simplifying the test. Overall the changes look good. Just a couple of minor suggestions inlined below/

File: test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java (hunk 3 above)
Comment (alexmenkov) [reply test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java]: Done.

File: src/hotspot/share/prims/jvmtiRedefineClasses.cpp (hunk 4 above)
Comment (alexmenkov) [reply src/hotspot/share/prims/jvmtiRedefineClasses.cpp]: Fixed. Also added explicit cast int -> u2

File: test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        originalClassBytes = Files.readAllBytes(origClassFile.toPath());
+        log("Read " + originalClassBytes.length + " bytes.");
+    }
+
+    private void log(Object o) {
+        System.out.println(String.valueOf(o));
+    }
+
+    // Prints and verifies MethodParameters attribute using reflection.
+    private void verifyMethodParams(boolean expectedPresent, String... expectedNames) throws Throwable {
```
Comment (sspitsyn) [inline test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java:101]: Just a side comment, no intention to insist...
I think, a generalization with the `expectedPresent` parameter adds a little bit of complexity.
It feels that refactoring it into two methods - one for expected and another for non-expected would be more clear.

**Review by sspitsyn — APPROVED** [review]

File: test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java (hunk 5 above)
Comment (alexmenkov) [reply test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java]: Fixed

**Review by plummercj — APPROVED** [review]

File: test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+        log("Reading test class from " + origClassFile);
+        originalClassBytes = Files.readAllBytes(origClassFile.toPath());
+        log("Read " + originalClassBytes.length + " bytes.");
+    }
+
+    private void log(Object o) {
+        System.out.println(String.valueOf(o));
+    }
+
+    private Parameter[] getTargetClassParameters() throws ClassNotFoundException {
```
Comment (sspitsyn) [inline test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java:100]: Should this be named getTargetMethodParameters instead?

**Review by sspitsyn — APPROVED** [review]
The update looks good to me.
I've inlined just one minor comment about naming.
Thanks,
Serguei

**Comment by fmeum** [comment]
@alexmenkov We noticed that the patch applies cleanly to JDK 17, but hasn't been backported yet. Our Java fuzzer [Jazzer](https://github.com/CodeIntelligenceTesting/jazzer) relies on class retransformation and, without this patch, is often unable to fuzz Spring applications. 

Would it be possible to backport this patch? If there is anything we can do to help let us know (we don't have contributor access).

**Pull request closed without merging**